File: src/de/saschahlusiak/freebloks/view/opengl/BoardRenderer.java
package de.saschahlusiak.freebloks.view.opengl;

import javax.microedition.khronos.opengles.GL10;
import de.saschahlusiak.freebloks.model.Player;
import de.saschahlusiak.freebloks.model.Spiel;
import de.saschahlusiak.freebloks.model.Stone;

public class BoardRenderer {
	final static float bevel_size = 0.18f;
	final static float bevel_height = 0.25f;
	final static float border_bottom = -1.2f;
	final public static float stone_size = 0.45f;
	
	SimpleModel field;
	SimpleModel border;
	SimpleModel stone;
	
    final float r1 = stone_size;
    final float r2 = stone_size - bevel_size;
    final float y1 = -bevel_height;
    final float y2 = 0.0f;
    
    Spiel spiel;


	private void initField() {
		field = new SimpleModel(8, 10);
		/* bottom, inner */
		field.addVertex(-r2, y1, +r2,  0, 1, 0,  0, 0);
		field.addVertex(+r2, y1, +r2,  0, 1, 0,  0, 0);
		field.addVertex(+r2, y1, -r2,  0, 1, 0,  0, 0);
		field.addVertex(-r2, y1, -r2,  0, 1, 0,  0, 0);

		/* top, outer */
		field.addVertex(-r1, y2, +r1, +1, 1, -1,  0, 0);
		field.addVertex(+r1, y2, +r1, -1, 1, -1,  0, 0);
		field.addVertex(+r1, y2, -r1, -1, 1, +1,  0, 0);
		field.addVertex(-r1, y2, -r1, +1, 1, +1,  0, 0);
		
		field.addIndex(0, 1, 2);
		field.addIndex(0, 2, 3);
		field.addIndex(0, 5, 1);
		field.addIndex(0, 4, 5);
		field.addIndex(1, 5, 6);
		field.addIndex(1, 6, 2);
		field.addIndex(2, 6, 7);
		field.addIndex(2, 7, 3);
		field.addIndex(3, 7, 4);
		field.addIndex(3, 4, 0);
		
		field.commit();
	}

	private void initStone() {
		stone = new SimpleModel(12, 20);
		/* top, inner */
		stone.addVertex(-r2, -y1, +r2,  0, 1, 0,  0, 0);
		stone.addVertex(+r2, -y1, +r2,  0, 1, 0,  0, 0);
		stone.addVertex(+r2, -y1, -r2,  0, 1, 0,  0, 0);
		stone.addVertex(-r2, -y1, -r2,  0, 1, 0,  0, 0);

		/* middle, outer */
		stone.addVertex(-r1, y2, +r1, -1, 0, +1,  0, 0);
		stone.addVertex(+r1, y2, +r1, +1, 0, +1,  0, 0);
		stone.addVertex(+r1, y2, -r1, +1, 0, -1,  0, 0);
		stone.addVertex(-r1, y2, -r1, -1, 0, -1,  0, 0);

		/* bottom, inner */
		stone.addVertex(-r2, y1, +r2,  0, -1, 0,  0, 0);
		stone.addVertex(+r2, y1, +r2,  0, -1, 0,  0, 0);
		stone.addVertex(+r2, y1, -r2,  0, -1, 0,  0, 0);
		stone.addVertex(-r2, y1, -r2,  0, -1, 0,  0, 0);
		
		/* top */
		stone.addIndex(0, 1, 2);
		stone.addIndex(0, 2, 3);
		stone.addIndex(0, 5, 1);
		stone.addIndex(0, 4, 5);
		stone.addIndex(1, 5, 6);
		stone.addIndex(1, 6, 2);
		stone.addIndex(2, 6, 7);
		stone.addIndex(2, 7, 3);
		stone.addIndex(3, 7, 4);
		stone.addIndex(3, 4, 0);
		
		/* bottom */
		stone.addIndex(8, 10, 9);
		stone.addIndex(8, 11, 10);
		stone.addIndex(8, 9, 5);
		stone.addIndex(8, 5, 4);
		stone.addIndex(9, 6, 5);
		stone.addIndex(9, 10, 6);
		stone.addIndex(10, 7, 6);
		stone.addIndex(10, 11, 7);
		stone.addIndex(11, 4, 7);
		stone.addIndex(11, 8, 4);
		
		stone.commit();
	}
	
	private void initBorder() {
		border = new SimpleModel(4, 2);
		float w;
	    
	    if (spiel == null)
			w = 5.0f;
		else
			w = stone_size * spiel.m_field_size_x;

		border.addVertex(w, 0, w, 0, 0, 1, 0, 0);
		border.addVertex(-w, 0, w, 0, 0, 1, 0, 0);
		border.addVertex(-w, border_bottom, w, 0, 0, 1, 0, 0);
		border.addVertex(w, border_bottom, w, 0, 0, 1, 0, 0);
		border.addIndex(0, 1, 2);
		border.addIndex(0, 2, 3);
		border.commit();
	}
	
	BoardRenderer(Spiel spiel) {
		this.spiel = spiel;
		initField();
		initBorder();
		initStone();
	}
	
	final float board_diffuse_normal[] = {0.52f,0.52f,0.50f,1.0f};
	final float board_diffuse_available[] = {0.5f,0.65f,0.5f,1.0f};
	final float board_specular[] = {0.27f,0.25f,0.25f,1.0f};
	final float board_shininess[] = {35.0f};

	public void renderBoard(GL10 gl, int currentPlayer) {
		gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_AMBIENT_AND_DIFFUSE, board_diffuse_normal, 0);
		gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SPECULAR, board_specular, 0);
		gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SHININESS, board_shininess, 0);
	 
	    gl.glVertexPointer(3, GL10.GL_FLOAT, 0, field.getVertexBuffer());
	    gl.glNormalPointer(GL10.GL_FLOAT, 0, field.getNormalBuffer());

	    if (spiel == null)
	    	return;
	    
	    gl.glPushMatrix();
	    gl.glTranslatef(-stone_size * (float)(spiel.m_field_size_x - 1), 0, -stone_size * (float)(spiel.m_field_size_x - 1) );
	    for (int y = 0; y < spiel.m_field_size_y; y++) {
	    	int x;
	    	for (x = 0; x < spiel.m_field_size_y; x++) {
	    		if (currentPlayer >= 0 && spiel.get_game_field(currentPlayer, y, x) == Stone.FIELD_ALLOWED)
	    			gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_AMBIENT_AND_DIFFUSE, board_diffuse_available, 0);
	    		else
	    			gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_AMBIENT_AND_DIFFUSE, board_diffuse_normal, 0);

	    		field.drawElements(gl);
	    		gl.glTranslatef(stone_size * 2.0f, 0, 0);
	    	}
	    	gl.glTranslatef(- x * stone_size * 2.0f, 0, stone_size * 2.0f);
	    }
	    gl.glPopMatrix();
		gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_AMBIENT_AND_DIFFUSE, board_diffuse_normal, 0);

	    gl.glVertexPointer(3, GL10.GL_FLOAT, 0, border.getVertexBuffer());
	    gl.glNormalPointer(GL10.GL_FLOAT, 0, border.getNormalBuffer());
	    for (int i = 0; i < 4; i++) {
	    	border.drawElements(gl);
	    	gl.glRotatef(90, 0, 1, 0);
	    }
	}
	
	final float stone_red[]={0.75f, 0, 0, 0};
	final float stone_blue[]={0.0f, 0.05f, 0.8f, 0};
	final float stone_green[]={0.0f, 0.75f, 0, 0};
	final float stone_yellow[]={0.75f, 0.75f, 0, 0};
	final float stone_white[]={0.7f, 0.7f, 0.7f, 0};
	final float stone_color_a[][] = { stone_white, stone_blue, stone_yellow, stone_red, stone_green };
	final float stone_specular[]={0.4f,0.4f,0.4f,1.0f};
	final float stone_shininess[]={40.0f};

	public void renderStone(GL10 gl, int color, float alpha) {
		final float c[] = stone_color_a[color + 1];
		c[3] = alpha;
		
		gl.glEnable(GL10.GL_BLEND);
		gl.glBlendFunc(GL10.GL_SRC_ALPHA, GL10.GL_ONE_MINUS_SRC_ALPHA);

		gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_AMBIENT_AND_DIFFUSE, c, 0);
		gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SPECULAR, stone_specular, 0);
		gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SHININESS, stone_shininess, 0);
		
	 
	    gl.glVertexPointer(3, GL10.GL_FLOAT, 0, stone.getVertexBuffer());
	    gl.glNormalPointer(GL10.GL_FLOAT, 0, stone.getNormalBuffer());
	    
	    stone.drawElements(gl, GL10.GL_TRIANGLES);
    	
    	gl.glDisable(GL10.GL_BLEND);
	}
	
	public void renderPlayerStones(GL10 gl, int player, float wheelAngle, int highlightStone, Stone currentStone) {
		final float da = 17.0f;
		float angle = wheelAngle + 9.5f * 0.5f * da;
		
		if (spiel == null)
			return;
		
		
		gl.glPushMatrix();
		gl.glTranslatef(0, -stone_size * 27.0f, 0);
		gl.glRotatef(wheelAngle, 0, 0, 1);
		gl.glTranslatef(-stone_size * 5.1f * 6.5f, 0, stone_size * (spiel.m_field_size_x + 8));
		gl.glRotatef(9.5f * 0.5f * da, 0, 0, 1);
		gl.glPushMatrix();
//		gl.glScalef(0.6f, 0.6f, 0.6f);
		Player p = spiel.get_player(player);
		for (int i = 0; i < Stone.STONE_COUNT_ALL_SHAPES; i++) {
			Stone s = p.get_stone(i);
			float alpha = 1.0f;
			while (angle < -180f)
				angle += 360.0f;
			while (angle > 180.0f)
				angle -= 360.0f;
			
			alpha = 1.0f / (1.0f + Math.abs(angle) / 47.0f);
			
			if (s.get_available() - ((s == currentStone) ? 1 : 0) > 0) {
				gl.glRotatef(90 * player, 0, 1, 0);
				renderPlayerStone(gl, (i == highlightStone) ? -1 : player, s, alpha);
				gl.glRotatef(-90 * player, 0, 1, 0);
			}
			
			gl.glTranslatef(stone_size * 2.0f * 5.1f, 0, 0);
			if (i % 11 == 10) {
				gl.glPopMatrix();
				gl.glTranslatef(0, 0, stone_size * 2.0f * 5.1f);
				angle = wheelAngle + 9.5f * 0.5f * da;
				gl.glPushMatrix();
			} else {
				gl.glRotatef(-da, 0, 0, 1);
				angle -= da;
			}
		}
		
		gl.glPopMatrix();
		gl.glPopMatrix();
	}
	
	public void renderPlayerStone(GL10 gl, int player, Stone stone, float alpha) {
		int i;
		gl.glTranslatef(-stone.get_stone_size() * stone_size, 0, -stone.get_stone_size() * stone_size);
		for (i = 0; i < stone.get_stone_size(); i++) {
			int j;
			for (j = 0; j < stone.get_stone_size(); j++) {				
				if (stone.get_stone_field(i,  j) != Stone.STONE_FIELD_FREE)
					renderStone(gl, player, 0.65f * alpha);
				gl.glTranslatef(stone_size * 2.0f, 0, 0);
			}
			gl.glTranslatef(-j*stone_size * 2.0f, 0, stone_size * 2.0f);
		}
		gl.glTranslatef(0, 0, -i * stone_size * 2.0f);
		gl.glTranslatef(stone.get_stone_size() * stone_size, 0, stone.get_stone_size() * stone_size);
	}
	
	public void renderPlayerStone(GL10 gl, int player, Stone stone, float alpha, int x, int y) {
	    gl.glTranslatef(
	    		-stone_size * (float)(spiel.m_field_size_x - 1) + stone_size * 2.0f * (float)x,
	    		0,
	    		+stone_size * (float)(spiel.m_field_size_x - 1) - stone_size * 2.0f * (float)y);
	    
		for (int i = 0; i < stone.get_stone_size(); i++) {
			int j;
			for (j = 0; j < stone.get_stone_size(); j++) {				
				if (stone.get_stone_field(i,  j) != Stone.STONE_FIELD_FREE)
					renderStone(gl, player, alpha);
				gl.glTranslatef(stone_size * 2.0f, 0, 0);
			}
			gl.glTranslatef(-j*stone_size * 2.0f, 0, stone_size * 2.0f);
		}
	}
}

File: app/src/main/kotlin/de/saschahlusiak/freebloks/view/BoardRenderer.kt
package de.saschahlusiak.freebloks.view

import android.content.res.Resources
import android.graphics.BitmapFactory
import android.opengl.GLUtils
import de.saschahlusiak.freebloks.R
import de.saschahlusiak.freebloks.ktx.KTX
import de.saschahlusiak.freebloks.model.Board
import de.saschahlusiak.freebloks.model.Orientation
import de.saschahlusiak.freebloks.model.Shape
import de.saschahlusiak.freebloks.model.StoneColor
import de.saschahlusiak.freebloks.theme.ColorThemes
import de.saschahlusiak.freebloks.theme.Theme
import javax.microedition.khronos.opengles.GL10
import javax.microedition.khronos.opengles.GL11

/**
 * Renders the content of the board, including the board itself and the player's stones, or any stone, really.
 */
class BoardRenderer(private val resources: Resources) {
    companion object {
        const val stoneSize = 0.45f
        const val defaultStoneAlpha = 0.75f

        private const val bevelSize = 0.18f
        private const val bevelHeight = 0.25f
        private const val borderBottom = -1.1f
        private const val borderTop = 0.4f
        private const val borderWidth = 0.5f

        private const val r1 = stoneSize
        private const val r2 = stoneSize - bevelSize
        private const val y1 = -bevelHeight
        private const val y2 = 0.0f

        private val materialBoardDiffuseSeed = floatArrayOf(0.45f, 0.85f, 0.6f, 0.75f)
        private val materialBoardSpecular = floatArrayOf(0.25f, 0.25f, 0.25f, 1.0f)
        private val materialBoardShininess = floatArrayOf(35.0f)
        val materialStoneSpecular = floatArrayOf(0.3f, 0.3f, 0.3f, 1.0f)
        val materialStoneShininess = floatArrayOf(30.0f)
        private val materialBlack = floatArrayOf(0f, 0f, 0f, 1f)
    }

    private val field: SimpleModel
    private var border: SimpleModel
    val stone: SimpleModel
    private val shadow: SimpleModel

    private var theme: Theme = ColorThemes.Black
    private var texture = IntArray(2)
    private val tmp = FloatArray(4)
    private var valid = false
    private var hasTexture = false
    private val rgba = FloatArray(4)

    init {
        field = buildSingleBoardFieldModel()
        border = buildBorderModel(Board.DEFAULT_BOARD_SIZE)
        stone = buildStoneModel()
        shadow = buildStoneShadow()
    }

    /**
     * Applies a new theme to the board
     */
    fun setTheme(theme: Theme) {
        this.theme = theme
        valid = false
    }

    /**
     * Builds a SimpleModel for a single field of the board, which is half a stone and is "deep".
     */
    private fun buildSingleBoardFieldModel(): SimpleModel {
        return SimpleModel(8, 10, true).apply {
            /* bottom, inner */
            addVertex(-r2, y1, +r2, 0f, 1f, 0f, -r2, r2)
            addVertex(+r2, y1, +r2, 0f, 1f, 0f, r2, r2)
            addVertex(+r2, y1, -r2, 0f, 1f, 0f, r2, -r2)
            addVertex(-r2, y1, -r2, 0f, 1f, 0f, -r2, -r2)

            /* top, outer */
            addVertex(-r1, y2, +r1, +1f, 1f, -1f, -r1, r1)
            addVertex(+r1, y2, +r1, -1f, 1f, -1f, r1, r1)
            addVertex(+r1, y2, -r1, -1f, 1f, +1f, r1, -r1)
            addVertex(-r1, y2, -r1, +1f, 1f, +1f, -r1, -r1)

            addIndex(0, 1, 2)
            addIndex(0, 2, 3)
            addIndex(0, 5, 1)
            addIndex(0, 4, 5)
            addIndex(1, 5, 6)
            addIndex(1, 6, 2)
            addIndex(2, 6, 7)
            addIndex(2, 7, 3)
            addIndex(3, 7, 4)
            addIndex(3, 4, 0)

            commit()
        }
    }

    /**
     * Builds the planar shadow quad for a single stone
     */
    private fun buildStoneShadow(): SimpleModel {
        return SimpleModel(4, 2, false).apply {
            addVertex(-r1, 0f, +r1, 0f, 1f, 0f, 0f, 0f)
            addVertex(+r1, 0f, +r1, 0f, 1f, 0f, 1f, 0f)
            addVertex(+r1, 0f, -r1, 0f, 1f, 0f, 1f, 1f)
            addVertex(-r1, 0f, -r1, 0f, 1f, 0f, 0f, 1f)
            addIndex(0, 1, 2)
            addIndex(0, 2, 3)
            commit()
        }
    }

    /**
     * Builds the model of a full player's stone, including top and bottom
     */
    private fun buildStoneModel(): SimpleModel {
        return SimpleModel(12, 20, true).apply {
            /* top, inner */
            addVertex(-r2, -y1, +r2, 0f, 1f, 0f, 0f, 0f)
            addVertex(+r2, -y1, +r2, 0f, 1f, 0f, 0f, 0f)
            addVertex(+r2, -y1, -r2, 0f, 1f, 0f, 0f, 0f)
            addVertex(-r2, -y1, -r2, 0f, 1f, 0f, 0f, 0f)

            /* middle, outer */
            addVertex(-r1, y2, +r1, -1f, 0f, +1f, 0f, 0f)
            addVertex(+r1, y2, +r1, +1f, 0f, +1f, 0f, 0f)
            addVertex(+r1, y2, -r1, +1f, 0f, -1f, 0f, 0f)
            addVertex(-r1, y2, -r1, -1f, 0f, -1f, 0f, 0f)

            /* bottom, inner */
            addVertex(-r2, y1, +r2, 0f, -1f, 0f, 0f, 0f)
            addVertex(+r2, y1, +r2, 0f, -1f, 0f, 0f, 0f)
            addVertex(+r2, y1, -r2, 0f, -1f, 0f, 0f, 0f)
            addVertex(-r2, y1, -r2, 0f, -1f, 0f, 0f, 0f)

            /* top */
            addIndex(0, 1, 2)
            addIndex(0, 2, 3)
            addIndex(0, 5, 1)
            addIndex(0, 4, 5)
            addIndex(1, 5, 6)
            addIndex(1, 6, 2)
            addIndex(2, 6, 7)
            addIndex(2, 7, 3)
            addIndex(3, 7, 4)
            addIndex(3, 4, 0)

            /* bottom */
            addIndex(8, 10, 9)
            addIndex(8, 11, 10)
            addIndex(8, 9, 5)
            addIndex(8, 5, 4)
            addIndex(9, 6, 5)
            addIndex(9, 10, 6)
            addIndex(10, 7, 6)
            addIndex(10, 11, 7)
            addIndex(11, 4, 7)
            addIndex(11, 8, 4)
            commit()
        }
    }

    /**
     * Builds the square model of the border of the field with the given size
     */
    private fun buildBorderModel(boardSize: Int): SimpleModel {
        return SimpleModel(12, 6, true).apply {
            val w1 = stoneSize * boardSize
            val w2 = w1 + borderWidth

            /* front */addVertex(w2, borderTop, w2, 0f, 0f, 1f, w2, borderTop)
            addVertex(-w2, borderTop, w2, 0f, 0f, 1f, -w2, borderTop)
            addVertex(-w2, borderBottom, w2, 0f, 0f, 1f, -w2, borderBottom)
            addVertex(w2, borderBottom, w2, 0f, 0f, 1f, w2, borderBottom)

            /* top */addVertex(w2, borderTop, w2, 0f, 1f, 0f, w2, w2)
            addVertex(-w2, borderTop, w2, 0f, 1f, 0f, -w2, w2)
            addVertex(-w1, borderTop, w1, 0f, 1f, 0f, -w1, w1)
            addVertex(w1, borderTop, w1, 0f, 1f, 0f, w1, w1)

            /* inner */addVertex(w1, borderTop, w1, 0f, 0f, -1f, w1, borderTop)
            addVertex(-w1, borderTop, w1, 0f, 0f, -1f, -w1, borderTop)
            addVertex(-w1, 0f, w1, 0f, 0f, -1f, -w1, w1)
            addVertex(w1, 0f, w1, 0f, 0f, -1f, w1, w1)
            addIndex(0, 1, 2)
            addIndex(0, 2, 3)
            addIndex(7, 6, 4)
            addIndex(4, 6, 5)
            addIndex(9, 8, 10)
            addIndex(8, 11, 10)

            commit()
        }
    }

    fun setBoardSize(boardSize: Int) {
        this.border = buildBorderModel(boardSize)
    }

    fun onSurfaceChanged(gl: GL11) {
        stone.invalidateBuffers(gl)
        field.invalidateBuffers(gl)
        border.invalidateBuffers(gl)
        shadow.invalidateBuffers(gl)

        // wood texture
        gl.glGenTextures(texture.size, texture, 0)
        valid = false

        // shadow texture
        val bitmap = BitmapFactory.decodeResource(resources, R.drawable.stone_shadow)
        gl.glBindTexture(GL10.GL_TEXTURE_2D, texture[1])
        gl.glTexParameterx(GL10.GL_TEXTURE_2D, GL10.GL_TEXTURE_MIN_FILTER, GL10.GL_LINEAR_MIPMAP_NEAREST)
        gl.glTexParameterf(GL11.GL_TEXTURE_2D, GL11.GL_GENERATE_MIPMAP, GL11.GL_TRUE.toFloat())
        gl.glTexParameterx(GL10.GL_TEXTURE_2D, GL10.GL_TEXTURE_MAG_FILTER, GL10.GL_LINEAR)
        GLUtils.texImage2D(GL10.GL_TEXTURE_2D, 0, bitmap, 0)
    }

    private fun updateTexture(gl: GL11) {
        val asset = theme.asset
        texture = IntArray(2).also { texture ->
            gl.glGenTextures(2, texture, 0)
        }

        if (theme.isResource && asset != null) {
            hasTexture = true
            gl.glBindTexture(GL10.GL_TEXTURE_2D, texture[0])

            gl.glTexParameterx(GL10.GL_TEXTURE_2D, GL10.GL_TEXTURE_MIN_FILTER, GL10.GL_LINEAR_MIPMAP_LINEAR)
            gl.glTexParameterf(GL11.GL_TEXTURE_2D, GL11.GL_GENERATE_MIPMAP, GL11.GL_TRUE.toFloat())
            gl.glTexParameterx(GL10.GL_TEXTURE_2D, GL10.GL_TEXTURE_MAG_FILTER, GL10.GL_LINEAR)

            KTX.loadKTXTexture(resources.assets, asset)

            theme.getColor(resources, rgba)
            rgba[3] = 1.0f
        } else {
            hasTexture = false
            theme.getColor(resources, rgba)
            rgba[3] = 1.0f
        }

        gl.glBindTexture(GL10.GL_TEXTURE_2D, 0)
    }

    /**
     * Render the border and fields of the board, optionally with seeds
     *
     * @param gl GL11
     * @param board the board to render
     * @param seedsPlayer the number of the player whose "seeds" to render or -1 if none
     */
    fun renderBoard(gl: GL11, board: Board, seedsPlayer: Int) {
        if (!valid) {
            updateTexture(gl)
            valid = true
        }

        val textureScale = 0.12f
        val textureRotation = 1.0f

        val materialBoardDiffuse = rgba

        gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SPECULAR, materialBoardSpecular, 0)
        gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SHININESS, materialBoardShininess, 0)
        gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_AMBIENT_AND_DIFFUSE, materialBoardDiffuse, 0)

        if (hasTexture) {
            gl.glEnable(GL10.GL_TEXTURE_2D)
            gl.glBindTexture(GL10.GL_TEXTURE_2D, texture[0])
        }
        field.bindBuffers(gl)

        gl.glMatrixMode(GL10.GL_TEXTURE)
        gl.glLoadIdentity()
        gl.glScalef(textureScale, textureScale, 1f)
        gl.glRotatef(textureRotation, 0f, 0f, 1f)
        gl.glPushMatrix()

        gl.glMatrixMode(GL10.GL_MODELVIEW)

        val w = board.width
        val h = board.height

        gl.glPushMatrix()
        gl.glTranslatef(-stoneSize * (w - 1).toFloat(), 0f, -stoneSize * (h - 1).toFloat())

        gl.glMaterialfv(
            GL10.GL_FRONT_AND_BACK,
            GL10.GL_AMBIENT_AND_DIFFUSE,
            materialBoardDiffuse,
            0
        )

        for (y in 0 until h) {
            var x = 0
            while (x < w) {
                field.drawElements(gl, GL10.GL_TRIANGLES)

                if (seedsPlayer >= 0 && board.getFieldStatus(seedsPlayer, y, x) == Board.FIELD_ALLOWED) {
                    gl.glDisable(GL10.GL_TEXTURE_2D)
                    gl.glEnable(GL10.GL_BLEND)
                    gl.glBlendFunc(GL10.GL_SRC_ALPHA, GL10.GL_ONE_MINUS_SRC_ALPHA)

                    gl.glMaterialfv(
                        GL10.GL_FRONT_AND_BACK,
                        GL10.GL_AMBIENT_AND_DIFFUSE,
                        materialBoardDiffuseSeed,
                        0
                    )

                    field.drawElements(gl, GL10.GL_TRIANGLES)

                    gl.glMaterialfv(
                        GL10.GL_FRONT_AND_BACK,
                        GL10.GL_AMBIENT_AND_DIFFUSE,
                        materialBoardDiffuse,
                        0
                    )

                    if (hasTexture) {
                        gl.glEnable(GL10.GL_TEXTURE_2D)
                    }
                    gl.glDisable(GL10.GL_BLEND)
                }

                gl.glMatrixMode(GL10.GL_TEXTURE)
                gl.glTranslatef(stoneSize * 4.0f, 0f, 0f)

                gl.glMatrixMode(GL10.GL_MODELVIEW)
                gl.glTranslatef(stoneSize * 2.0f, 0f, 0f)

                x++
            }
            gl.glMatrixMode(GL10.GL_TEXTURE)
            gl.glTranslatef(-x * stoneSize * 4.0f, stoneSize *4.0f, 0f)

            gl.glMatrixMode(GL10.GL_MODELVIEW)
            gl.glTranslatef(-x * stoneSize * 2.0f, 0f, stoneSize * 2.0f)
        }

        gl.glMatrixMode(GL10.GL_TEXTURE)
        gl.glPopMatrix()

        gl.glMatrixMode(GL10.GL_MODELVIEW)
        gl.glPopMatrix()

        gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_AMBIENT_AND_DIFFUSE, materialBoardDiffuse, 0)

        border.bindBuffers(gl)

        /* we want the border in the depth buffer so it can cover the stones rendered later */
        gl.glEnable(GL10.GL_DEPTH_TEST)
        for (i in 0..3) {
            border.drawElements(gl, GL10.GL_TRIANGLES)
            gl.glRotatef(90f, 0f, 1f, 0f)
        }

        gl.glMatrixMode(GL10.GL_TEXTURE)
        gl.glLoadIdentity()
        gl.glMatrixMode(GL10.GL_MODELVIEW)
        gl.glDisable(GL10.GL_TEXTURE_2D)
    }

    /**
     * Render a single stone with the given color and alpha
     *
     * The [stone] model MUST be bound before calling this.
     *
     * This is ONLY used when rendering non-effected player stones of the field
     */
    fun renderSingleStone(gl: GL11, color: StoneColor, alpha: Float) {
        val c = color.stoneColor
        tmp[0] = c[0] * alpha
        tmp[1] = c[1] * alpha
        tmp[2] = c[2] * alpha
        tmp[3] = alpha
        gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_AMBIENT_AND_DIFFUSE, tmp, 0)
        stone.drawElements(gl, GL10.GL_TRIANGLES)
    }

    /**
     * Render the shadow of the shape at the current position.
     */
    fun renderShapeShadow(gl: GL11, color: StoneColor, shape: Shape, orientation: Orientation, alpha: Float) {
        val c = color.shadowColor
        tmp[0] = c[0] * alpha
        tmp[1] = c[1] * alpha
        tmp[2] = c[2] * alpha
        tmp[3] = alpha

        gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_AMBIENT_AND_DIFFUSE, tmp, 0)
        gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SPECULAR, materialBlack, 0)
        gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SHININESS, materialBlack, 0)

        shadow.bindBuffers(gl)

        gl.glEnable(GL10.GL_BLEND)
        gl.glBlendFunc(GL10.GL_ONE, GL10.GL_ONE_MINUS_SRC_ALPHA)
        gl.glEnable(GL10.GL_TEXTURE_2D)
        gl.glBindTexture(GL10.GL_TEXTURE_2D, texture[1])

        for (i in 0 until shape.size) {
            var j = 0
            while (j < shape.size) {
                if (shape.isStone(j, i, orientation))
                    shadow.drawElements(gl, GL10.GL_TRIANGLES)

                gl.glTranslatef(stoneSize * 2.0f, 0f, 0f)
                j++
            }
            gl.glTranslatef(-j * stoneSize * 2.0f, 0f, stoneSize * 2.0f)
        }
        gl.glDisable(GL10.GL_TEXTURE_2D)
        gl.glDisable(GL10.GL_BLEND)
    }

    fun renderShapeShadow(
        gl: GL11,
        shape: Shape,
        color: StoneColor,
        orientation: Orientation,
        height: Float,
        ang: Float, ax: Float, ay: Float, az: Float,
        lightAngle: Float,
        alpha: Float,
        scale: Float
    ) {
        val offset = shape.size.toFloat() - 1.0f
        val heightAlphaMultiplier = 0.80f - height / 16.0f
        if (heightAlphaMultiplier < 0.0f) return

        gl.glRotatef(-lightAngle, 0f, 1f, 0f)
        gl.glTranslatef(2.5f * height * 0.08f, 0f, 2.0f * height * 0.08f)
        gl.glRotatef(lightAngle, 0f, 1f, 0f)
        gl.glTranslatef(
            stoneSize * offset, 0f,
            stoneSize * offset
        )
        gl.glScalef(scale, 0.01f, scale)
        gl.glRotatef(ang, ax, ay, az)
        gl.glScalef(1.0f + height / 16.0f, 1f, 1.0f + height / 16.0f)
        gl.glTranslatef(
            -stoneSize * offset, 0f,
            -stoneSize * offset
        )

        renderShapeShadow(gl, color, shape, orientation, heightAlphaMultiplier * alpha)
    }

    /**
     * Renders the full shape at the current position and angle.
     */
    fun renderShape(gl: GL11, color: StoneColor, shape: Shape, orientation: Orientation, alpha: Float) {
        val c = color.stoneColor
        tmp[0] = c[0] * alpha
        tmp[1] = c[1] * alpha
        tmp[2] = c[2] * alpha
        tmp[3] = alpha

        gl.glEnable(GL10.GL_BLEND)
        gl.glBlendFunc(GL10.GL_ONE, GL10.GL_ONE_MINUS_SRC_ALPHA)
        gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_AMBIENT_AND_DIFFUSE, tmp, 0)
        gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SPECULAR, materialStoneSpecular, 0)
        gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SHININESS, materialStoneShininess, 0)
        stone.bindBuffers(gl)

        var i = 0
        while (i < shape.size) {
            var j = 0
            while (j < shape.size) {
                if (shape.isStone(j, i, orientation))
                    stone.drawElements(gl, GL10.GL_TRIANGLES)
                gl.glTranslatef(stoneSize * 2.0f, 0f, 0f)
                j++
            }
            gl.glTranslatef(-j * stoneSize * 2.0f, 0f, stoneSize * 2.0f)
            i++
        }
        gl.glDisable(GL10.GL_BLEND)
    }
}